### PR TITLE
Allow to specify the license location via directory argument

### PR DESCRIPTION
**Pull request by dgdavid**
## Problem

After changes made in https://github.com/yast/yast-firstboot/pull/67, it seems that the `Y2Firstboot::Clients::Licenses` client cannot be used to display two (or more) licenses in two (or more) different _pages_. Instead, when the client found more than once valid path to licenses in the firstboot configuration file it will show them together in the same _page_.

* https://bugzilla.novell.com/show_bug.cgi?id=1154708

<details>
<summary>Click to show/hide screenshots</summary>

---
<p align="center"><em>Defined path to licenses in the firstboot configuration file</em></p>

Screenshot_20191213_033645

---

<p align="center"><em>Licenses are being shown together</em></p>

Screenshot_20191213_033330
</details>

## Even using the `directory` argument

```xml
<module>
  <label>SUSE License Agreement</label>
  <enabled config:type="boolean">true</enabled>
  <name>firstboot_licenses</name>
  <arguments>
    <directory>/path/to/SUSE/licenses/dir</directory>
  </arguments>
</module>
```
To use the `directory` argument as above has no effect because `directories` argument is also sent to the `InstLicense` installation client, which [gives precedence to the latter over the former](https://github.com/yast/yast-installation/blob/aa816f10b7fa748eb7aafb9ecbacfbf9b2bab718/src/lib/installation/clients/inst_license.rb#L83-L90).

## Solution

To choose the path(s) to be included in the `directories` argument sent to the `InstLicense` installation client, prioritizing `directory` module argument when present.

## Test

* Unit test updated.
* Also tested manually.

## Screenshots

<details>
<summary>Click to show/hide</summary>

---

<p align="center"><em>The firstboot configuration</em></p>

| /etc/sysconfig/firstboot | /etc/YaST2/firstboot.xml |
| - | - |
| Screenshot_20191213_035019 | Screenshot_20191213_034719 |

---

<p align="center"><em>openSUSE License Agreement</em></p>

Screenshot_20191213_032758

---

<p align="center"><em>Partner License Agreement</em></p>

Screenshot_20191213_032934
</details>

### CODE REVIEW

File: test/y2firstboot/clients/licenses_test.rb [hunk 1]
```diff
@@ -44,53 +45,71 @@
...
       let(:expected_arg) do
-        { "directories" => [firstboot_license_dir] }
+        { "directories" => [directory] }
       end
 
-      it "includes it as arg for InstLicense client" do
+      it "includes it as an argument for InstLicense client" do
         expect(Yast::WFM).to receive(:CallFunction)
-          .with("inst_license", array_including(hash_including(expected_arg)))
+          .with("inst_license", array_including(hash_including("directory" => directory)))
```
Comment (joseivanlopez) [inline test/y2firstboot/clients/licenses_test.rb]: NP: why not `expected_arg`?

Comment (dgdavid) [reply test/y2firstboot/clients/licenses_test.rb]: Because `expected_arg` is a leftover since I removed it just to be more explicit.

File: src/lib/y2firstboot/clients/licenses.rb [hunk 2]
```diff
@@ -53,7 +53,7 @@ def run
       def inst_license_args
         args = GetInstArgs.argmap
         args["action"] = refusal_action
-        args["directories"] = licenses_directories
+        args["directories"] = licenses_directories if args["directory"].to_s.empty?
```
Comment (joseivanlopez) [inline src/lib/y2firstboot/clients/licenses.rb]: As already commented offline, I would prefer to continue using `directories` parameter, but with the proper license paths for each case: directory path from the module or directory paths from the sysconfig file. This will allow to show several module licenses in the same page (in case someone requests that in the future).

File: test/y2firstboot/clients/licenses_test.rb (hunk 1 above)
Comment (joseivanlopez) [reply test/y2firstboot/clients/licenses_test.rb]: Just above I see this:

~~~
      let(:expected_arg) do
        { "directories" => [directory] }
      end
~~~

So we could use it if we adapt the code to always use `directories` key.

Comment (dgdavid) [reply test/y2firstboot/clients/licenses_test.rb]: > Just above I see this:
> ...

Yes, because I forgot to remove it. That's why said that it is intentionally to use the explicit expectation and the `expected_arg` is a leftover.

File: src/lib/y2firstboot/clients/licenses.rb (hunk 2 above)
Comment (dgdavid) [reply src/lib/y2firstboot/clients/licenses.rb]: Done. See the new commit.

File: src/lib/y2firstboot/clients/licenses.rb [hunk 3]
```diff
@@ -69,18 +72,35 @@ def refusal_action
...
+      # path given as "base_product_license_directory" global param through the control file.
+      #
+      # @return [Array<String>] license agreement paths
+      def directories
+        (module_license_directories + sysconfig_licenses_directories).compact.reject(&:empty?).uniq
+      end
+
+      # Directories defined by the module arguments
+      #
+      # NOTE: for the time being, it is only based in the "directory" argument, but it should be
```
[the remark this answers is not in the document]
Comment (joseivanlopez) [reply src/lib/y2firstboot/clients/licenses.rb]: NP: I would rephrase it a little:

NOTE: right now license module can only define a directory path, but it could be extended to support more than one (e.g., by a "directories" list).

File: src/lib/y2firstboot/clients/licenses.rb [hunk 4]
```diff
@@ -69,18 +72,35 @@ def refusal_action
...
+      # used also if the support for the "directories" argument is implemented.
       #
       # @return [Array<String>] license agreement paths
-      def licenses_directories
-        directories = [
+      def module_license_directories
+        [
+          args["directory"]
```
Comment (joseivanlopez) [inline src/lib/y2firstboot/clients/licenses.rb]: I would return an empty list if no directory is given:

```suggestion
          [ args["directory"] ].map(&:to_s).reject(&:empty?)
```

File: src/lib/y2firstboot/clients/licenses.rb [hunk 5]
```diff
@@ -69,18 +72,35 @@ def refusal_action
...
+      # Directories defined in the sysconfig firstboot file
+      #
+      # @return [Array<String>] license agreement paths
+      def sysconfig_licenses_directories
+        [
           Misc.SysconfigRead(path(".sysconfig.firstboot.FIRSTBOOT_LICENSE_DIR"), ""),
           Misc.SysconfigRead(path(".sysconfig.firstboot.FIRSTBOOT_NOVELL_LICENSE_DIR"), "")
         ]
-
-        directories.uniq.reject(&:empty?)
```
Comment (joseivanlopez) [inline src/lib/y2firstboot/clients/licenses.rb]: I would keep this line. I think the method should be in charge of filtering leftover entries (similar to `#module_license_directories`).

BTW, I would rename the method as `#sysconfig_license_directories` (to keep correspondence with the other method name `#module_license_directories`).

File: src/lib/y2firstboot/clients/licenses.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (dgdavid) [reply src/lib/y2firstboot/clients/licenses.rb]: > NP: "base in" or "base on" ?

This is a question for me or for @mvidner? ;)

_Based on_ [this answer](https://www.italki.com/question/213376), **based on** is more appropriated.

Thanks!

File: src/lib/y2firstboot/clients/licenses.rb (hunk 5 above)
Comment (dgdavid) [reply src/lib/y2firstboot/clients/licenses.rb]: Actually, I miss an `s` in the `module_license_directories`, but I'm going to use your suggestion.
